src/evict/evict_page.c

```diff
@@ -675,46 +676,45 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
+     * and restore the original transaction's context (including snapshot) once we are finished.
+     */
 
     /*
      * TODO: We are deliberately not using a snapshot when checkpoint is active. This will ensure
      * that point-in-time checkpoints have a consistent version of data. Remove this condition once
      * fuzzy transaction ID based checkpoints work is merged.
      */
-    if (FLD_ISSET(evict_flags, WT_REC_EVICTION_THREAD) && !WT_IS_HS(S2BT(session)) &&
-      !conn->txn_global.checkpoint_running) {
-        /*
-         * We rely on the fact that the eviction threads are created with read committed isolation
-         * by default. If this fact doesn't hold anymore in future, we have to force isolation
-         * level.
-         */
-        WT_ASSERT(session, session->txn->isolation == WT_ISO_READ_COMMITTED);
-
-        /*
-         * Eviction threads do not need to pin anything in the cache. We have a exclusive lock for
-         * the page being evicted so we are sure that the page will always be there while it is
-         * being processed. Therefore, we use snapshot API that doesn't publish shared IDs to the
-         * outside world.
-         */
-        __wt_txn_bump_snapshot(session);
-
+    /*
+     * Avoid updating snapshots when application transactions are in final stages of commit or
+     * rollback as they have already released the snapshot. It otherwise makes it harder in later
+     * part of code to detect updates the belonged to last the running transaction from application
+     * thread.
+     */
+    use_snapshot_for_app_thread = !F_ISSET(session, WT_SESSION_INTERNAL) &&
+      !WT_IS_METADATA(session->dhandle) && WT_SESSION_TXN_SHARED(session)->id != WT_TXN_NONE;
+    use_snapshot_for_eviction_thread = FLD_ISSET(evict_flags, WT_REC_EVICTION_THREAD);
+
+    if (!conn->txn_global.checkpoint_running && !WT_IS_HS(S2BT(session)) &&
+      (use_snapshot_for_app_thread || use_snapshot_for_eviction_thread)) {
+        WT_RET(__wt_txn_save_and_update_snapshot(session, &old_state));
+        restore_snapshot = true;
         /*
          * Make sure once more that there is no checkpoint running. A new checkpoint might have
          * started between previous check and acquiring snapshot. If there is a checkpoint running,
          * release the checkpoint and fallback to global visibility checks.
          */
         if (conn->txn_global.checkpoint_running) {
-            __wt_txn_release_snapshot(session);
+            __wt_txn_restore_snapshot(session, &old_state);
             LF_SET(WT_REC_VISIBLE_ALL);
-        } else {
-            /*
-             * If we acquired a snapshot for eviction, force the isolation to ensure the snapshot
-             * isn't released when history store cursors are closed.
-             */
-            ++session->txn->forced_iso;
-            snapshot_acquired = true;
-        }
+            restore_snapshot = false;
+        } else if (use_snapshot_for_app_thread)
```

Comment:
possibly merge the two bools as mentioned above and use "! FLD_ISSET(evict_flags, WT_REC_EVICTION_THREAD);" here ?